### PR TITLE
Define interface for client.

### PR DESCRIPTION
**Pull request by chowchow316**
Basic design is: 
* Mixer client only provides cache and aggregation functions. 
* Transport is separated out from this client so different platforms can use this client with their different transports.
* Client interface is exactly the same as transport interface; easier to integrate.

Please help to review this interface.

### CODE REVIEW

File: include/client.h [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+==============================================================================*/
+
+#ifndef MIXERCLIENT_CLIENT_H
+#define MIXERCLIENT_CLIENT_H
+
+#include <functional>
+#include <memory>
+#include <string>
+
+#include "google/api/servicecontrol/v1/service_controller.pb.h"
```
Comment (qiwzhang) [inline include/client.h]: no need service_controller.pb
you may need mixer.api.pb.h

File: include/client.h [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+#include "google/protobuf/stubs/status.h"
+#include "options.h"
+
+namespace google {
+namespace mixer_client {
+
+// Defines a function prototype used when an asynchronous transport call
+// is completed.
+using TransportDoneFunc =
```
Comment (qiwzhang) [inline include/client.h]: Just call this DoneFunc,  use it for the one in line 69

File: include/client.h [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+  // Destructor
+  virtual ~MixerClient() {}
+
+  // A check call with provided per_request transport function.
+  // Only some special platforms may need to use this function.
+  // It allows caller to pass in a per_request transport function.
+  virtual void Check(const ::istio::mixer::v1::CheckRequest& check_request,
+                     ::istio::mixer::v1::CheckResponse* check_response,
+                     DoneCallback on_check_done,
+                     TransportCheckFunc check_transport) = 0;
```
Comment (qiwzhang) [inline include/client.h]: don't support per_request transport.   remove param 4.

File: include/client.h [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+  // A report call with provided per_request transport function.
+  // Only some special platforms may need to use this function.
+  // It allows callers to pass in a per_request transport function.
+  virtual void Report(const ::istio::mixer::v1::ReportResponse& report_request,
+                      ::istio::mixer::v1::ReportResponse* report_response,
+                      DoneCallback on_report_done,
+                      TransportReportFunc report_transport) = 0;
```
Comment (qiwzhang) [inline include/client.h]: remove para4.

File: include/client.h [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+};
+
+// Creates a MixerClient object.
+std::unique_ptr<MixerClient> CreateMixerClient(
+    const std::string& service_name, const std::string& service_config_id,
```
Comment (qiwzhang) [inline include/client.h]: no need service_name, and config_id for now.

File: include/options.h [hunk 6]
```diff
@@ -0,0 +1,83 @@
+/* Copyright 2016 Google Inc. All Rights Reserved.
```
Comment (qiwzhang) [inline include/options.h]: It seems that format for copyright header is wrong.

File: include/client.h (hunk 1 above)
Comment (chowchow316) [inline include/client.h]: Done. Will add necessary includes when building with cpp code.

File: include/client.h (hunk 5 above)
Comment (chowchow316) [inline include/client.h]: Done.

File: include/options.h (hunk 6 above)
Comment (chowchow316) [inline include/options.h]: Replaced with istio/proxy copyright format.

File: include/client.h (hunk 3 above)
Comment (chowchow316) [inline include/client.h]: Done.

File: include/client.h (hunk 4 above)
Comment (chowchow316) [inline include/client.h]: Done.

File: include/client.h (hunk 2 above)
Comment (chowchow316) [inline include/client.h]: Done.

**Comment by qiwzhang** [comment]
Could you add Quota method too so the interface is complete?  QuotaOptions can be similar to CheckOptions.

File: include/client.h [hunk 7]
```diff
@@ -0,0 +1,116 @@
+/* Copyright 2016 Google Inc. All Rights Reserved.
```
Comment (douglas-reid) [inline include/client.h]: nit: it is now 2017

File: include/options.h [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+#include <memory>
+
+namespace google {
+namespace mixer_client {
+
+// Options controlling check behavior.
+struct CheckOptions {
+  // Default constructor.
+  CheckOptions()
+      : num_entries(10000), flush_interval_ms(500), expiration_ms(1000) {}
```
Comment (douglas-reid) [inline include/options.h]: are these default options chosen from experience, or for any particular reason?  is it worth documenting this choice somewhere?

Comment (qiwzhang) [inline include/options.h]: These are based pass experience.  For example, num_entries determines how much memory will be used for cache.  flush_interval will be based on what kind of checks.  For ESP, check is only checking api_key,  flush_interval is 5 minutes. 

We could document these in the comment.

File: include/client.h [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+#include <functional>
+#include <memory>
+#include <string>
+
+#include "google/protobuf/stubs/status.h"
+#include "mixer/api/v1/service.pb.h"
+#include "options.h"
+
+namespace google {
```
Comment (geeknoid) [inline include/client.h]: Should this be istio instead of google?

File: include/client.h [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+  // Destructor
+  virtual ~MixerClient() {}
+
+  // The async call.
+  // on_check_done is called with the check status after cached
+  // check_response is returned in case of cache hit, otherwise called after
+  // check_response is returned from the Controller service.
+  //
+  // check_response must be alive until on_check_done is called.
+  virtual void Check(const ::istio::mixer::v1::CheckRequest& check_request,
```
Comment (geeknoid) [inline include/client.h]: This API structure doesn't provide any support for cancellation. Is that OK?

Comment (qiwzhang) [inline include/client.h]: Transport layer has timeout which can serve as cancellation. ESF and ESP have been using this kind of API.  So far we are OK.

File: include/client.h (hunk 9 above)
Comment (chowchow316) [inline include/client.h]: Updated.

File: include/client.h (hunk 7 above)
Comment (chowchow316) [inline include/client.h]: Updated.

**Pull request merged**